File: pari3i.py
def add_two_lists(l1_head, l2_head):
    dummy_head = Node(0)
    current = dummy_head
    p1, p2 = l1_head, l2_head
    carry = 0
    
    while p1 or p2 or carry:
        val1 = p1.data if p1 else 0
        val2 = p2.data if p2 else 0
        
        total_sum = val1 + val2 + carry
        carry = total_sum // 10
        new_digit = total_sum % 10
        
        current.next = Node(new_digit)
        current = current.next
        
        if p1: p1 = p1.next
        if p2: p2 = p2.next
        
    return dummy_head.next

l1 = SinglyLinkedList()
l1.insert_at_end(1)
l1.insert_at_end(2)
l1.insert_at_end(3)

l2 = SinglyLinkedList()
l2.insert_at_end(5)
l2.insert_at_end(6)
l2.insert_at_end(4)

print("List 1:", end=" ")
l1.print_list()
print("List 2:", end=" ")
l2.print_list()

sum_head = add_two_lists(l1.head, l2.head)
sum_list = SinglyLinkedList()
sum_list.head = sum_head
print("Sum List:", end=" ")
sum_list.print_list() 

File: pari3i.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* reverseList(struct Node* head) {
    struct Node *prev = NULL, *current = head, *next = NULL;
    while (current != NULL) {
        next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    return prev;
}

struct Node* addTwoLists(struct Node* first, struct Node* second) {
    first = reverseList(first);
    second = reverseList(second);

    int carry = 0;
    struct Node* sum_list_rev = NULL;
    while (first != NULL || second != NULL || carry != 0) {
        int sum = carry + (first ? first->data : 0) + (second ? second->data : 0);
        carry = sum / 10;
        sum = sum % 10;
        struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
        newNode->data = sum;
        newNode->next = sum_list_rev;
        sum_list_rev = newNode;

        if (first) first = first->next;
        if (second) second = second->next;
    }

    return sum_list_rev;
}
void printList(struct Node* head);
struct Node* createListFromArray(int arr[], int n);

int main() {
    int arr1[] = {4, 5};
    int arr2[] = {3, 4, 5};
    struct Node* l1 = createListFromArray(arr1, 2);
    struct Node* l2 = createListFromArray(arr2, 3);
    
    printf("List 1 (45): "); printList(l1);
    printf("List 2 (345): "); printList(l2);
    
    struct Node* sum_list = addTwoLists(l1, l2);
    
    printf("Sum List (390): ");
    printList(sum_list);
    
    return 0;
}
void printList(struct Node* head) { while (head) { printf("%d -> ", head->data); head = head->next; } printf("NULL\n"); }
struct Node* createListFromArray(int arr[], int n) { if (n == 0) return NULL; struct Node* head = (struct Node*)malloc(sizeof(struct Node)); head->data = arr[0]; struct Node* curr = head; for (int i = 1; i < n; i++) { curr->next = (struct Node*)malloc(sizeof(struct Node)); curr = curr->next; curr->data = arr[i]; } curr->next = NULL; return head; }
